apiserver/storage/storage.go

```diff
@@ -200,3 +201,64 @@ func (a *API) checkProviderTypeSupported(providerType storage.ProviderType) erro
 	}
 	return nil
 }
+
+var getVolumeManager = func(vs volume.VolumeState) volume.VolumeManager {
+	return volume.NewVolumeManager(vs)
+}
+
+func (a *API) ListVolumes(filter params.StorageVolumeFilter) (params.StorageVolumesResult, error) {
+	volumeManager := getVolumeManager(a.storage)
+
+	all, err := volumeManager.List()
+	if err != nil {
+		return params.StorageVolumesResult{}, err
+	}
+	results := []params.StorageDisk{}
+	// Convert to sets as easier to deal with
+	machineSet := set.NewStrings(filter.Machines...)
+	for _, disk := range all {
+		if one, k := filterDisk(machineSet, disk); k {
+			results = append(results, one)
+		}
+	}
+	return params.StorageVolumesResult{Disks: results}, nil
+}
+
+// filterDisk returns converted Disk and boolean indicating
+// if disk contains attachments that match filter
+func filterDisk(machineSet set.Strings, disk volume.Disk) (params.StorageDisk, bool) {
+	attachments := []params.StorageAttachment{}
+	for _, attachment := range disk.Attachments() {
+		if one, k := filterAttachment(machineSet, attachment); k {
+			attachments = append(attachments, one)
+		}
+	}
+	// it's possible that there will be no attachments on this disk
+	// that match the filter. This disk will be filtered out too then :D
```

Comment:
added a todo comment. as per irc, would be best to add this functionality when we can test for it.